lotman/lotman_ui.go

```diff
@@ -0,0 +1,573 @@
+		}
+		ownerFound = true
+		break
+	}
+
+	if !ownerFound {
+		return false, nil, errors.New("token not signed by any of the owners of any parent lot")
+	}
+
+	return true, &tok, nil
+}
+
+// Verify that a token received is a valid token. Upon verification, we set the lot's parents/owner to the
+// appropriate values. Returns true if the token is valid, false otherwise.
+func VerifyNewLotToken(lot *Lot, strToken string) (bool, error) {
+	tokenApproved := false
+
+	// Get the path associated with the lot (right now we assume/enforce a single path) and try
+	// to deduce a namespace prefix from that. We'll use that namespace prefix's issuer as the
+	// owner of the lot. If there is no associated isuer, we assign ownership to the federation
+
+	path := ((lot.Paths)[0]).Path
+	log.Debugf("Attempting to add lot for path: %s", path)
+	errMsg := make([]byte, 2048)
+	lots := unsafe.Pointer(nil)
+	// Pass a pointer to the first element of the slice to the C++ function.
+	ret := LotmanGetLotsFromDir(path, false, &lots, &errMsg)
+	if ret != 0 {
+		trimBuf(&errMsg)
+		return false, errors.Errorf("Error getting lot JSON: %s", string(errMsg))
+	}
+
+	goLots := cArrToGoArr(&lots)
+
+	// Check if goOut[0] is "default". Lotman should ALWAYS put something in this array. If it's empty, we have a problem.
+	if goLots[0] == "default" {
+		// We'll assign to the root lot. However, be careful here because we make an assumption -- under the hood,
+		// Lotman will return "default" if there's no other lot, or if the actual path is in some way assigned to
+		// the default lot. When we assign to parent to root here, we're assuming the former. This implies we should stay
+		// away from assigning paths to the default lot.
+		lot.Parents = []string{"root"}
+	} else {
+		// We found a different logical parent
+		lot.Parents = goLots
+	}
+
+	var tok jwt.Token
+	if len(lot.Parents) == 0 && lot.Parents[0] == "root" {
+		// We check that the token is signed by the federation
+		// First check for discovery URL and then for director URL, both of which should host the federation's pubkey
+		issuerUrl := param.Federation_DiscoveryUrl.GetString()
+		if issuerUrl == "" {
+			issuerUrl = param.Federation_DirectorUrl.GetString()
+			if issuerUrl == "" {
+				return false, errors.New("Federation discovery URL and director URL are not set")
+			}
+			log.Debugln("Federation discovery URL is not set, using director URL as lot token issuer")
+		}
+
+		kSet, err := server_utils.GetJWKSFromIssUrl(issuerUrl)
+		if err != nil {
+			return false, errors.Wrap(err, "Error getting JWKS from issuer URL")
+		}
+
+		tok, err = jwt.Parse([]byte(strToken), jwt.WithKeySet(*kSet), jwt.WithValidate(true))
+		if err != nil {
+			return false, errors.Wrap(err, "Error parsing token")
+		}
+	} else {
+
+		// For each parent that might be here, get all owners and check that the token is signed by one of them
+		owners := []string{}
+		internalOwners := []string{}
+		for _, parent := range lot.Parents {
```

Comment:
Do you want owners parents of parents to have the ability to manage lower lots? Because currently, this doesn't allow for that, but maybe you only want the assigned user of the lot (still testing out different terminology in the comments) to be able to create sub-lots.

If you do want the more, then I would again suggest a worklist algorithm. And since this would be the same as the one needed in lotman.go, I suggest creating a separate function to get all the ancestors.

Reply:
While this only gets the lot's immediate parents, we then recursively get all owners of those parents two lines down when we call `LotmanGetLotOwners`